### ISSUE TITLE
how to access input DOM in people picker? 

### ISSUE THREAD

**Issue by bhedayatian** [OfficeDev/office-ui-fabric-react#10175]
how to access input DOM in peoplepicker? I need to clear what user has entered and change the behaviour of people-picker when user past or type in the people-picker. can you please share some code in react ( not typescript)

**Comment by joschect** [comment]
@bhedayatian I think that the floating picker is probably a better fit for what you want to do. The people picker is an all in one solution, the floating picker gives you a lot more control over the input and what it's doing at any given moment.

Here's a [codesandbox](https://codesandbox.io/s/basic-floating-picker-w8je2) to help you get started with it!

**Comment by bhedayatian** [comment]
is floating picker part of fabric ui 5.39.1??

**Comment by joschect** [comment]
Yep! It is. Github's file search is pretty good and if you go to the 5.0 branch you can find it pretty easily. https://github.com/OfficeDev/office-ui-fabric-react/find/5.0.

Here's the file as well https://github.com/OfficeDev/office-ui-fabric-react/blob/5.0/packages/office-ui-fabric-react/src/components/FloatingPicker/BaseFloatingPicker.tsx

**Comment by JasonGore** [comment]
That said 5.39.1 is 2 major versions old and nearly 2 years old. If you can use a newer version of Fabric, you'll get the benefit of many fixes and updates. 😄

**Comment by bhedayatian** [comment]
Thank you for your quick reply.  The example was very helpful.
- is it possible to use inputElement property with NormalPeoplePicker? the look and behaviour of floating picker is a bit different than what i need. 

- I understand that in the github floating picker is part of branch 5, but when i installed  "office-ui-fabric-react": "5.39.1" as part of dependendies in the package.json, i cant find floating picker.

Thanks for your help

**Comment by joschect** [comment]
@bhedayatian It cannot be used with the NormalPeoplePicker. The floating picker can be styled so that it matches the people picker. For example take a look at the [ExtendedPeoplePicker](https://github.com/OfficeDev/office-ui-fabric-react/blob/5.0/packages/office-ui-fabric-react/src/components/ExtendedPicker/PeoplePicker/ExtendedPeoplePicker.tsx) which extends the [ExtendedPicker](https://github.com/OfficeDev/office-ui-fabric-react/blob/5.0/packages/office-ui-fabric-react/src/components/ExtendedPicker/PeoplePicker/ExtendedPeoplePicker.tsx).

**Comment by bhedayatian** [comment]
cool. i will try that. does this mean that using NormalPeoplePicker I can NOT access the input itself and clear the text user has entered in any way?

**Comment by bhedayatian** [comment]
how to add functions for onPaste and onKeyUp events? i added them as part of suggestionProps but didnt work. do you have an example?

**Comment by bhedayatian** [comment]
Can floating picker accept multi values (separated by comma for example)?

**Comment by joschect** [comment]
@bhedayatian Correct, the normal people picker is not supposed to enable access to the input itself.

I do not have an explicit example of the floating picker accepting multiple suggestions, but I know that it's been used to accomplish behavior like that. It really depends on how you're implemented search function works. Since the floating picker takes a list of suggestions in you can control whatever you want to be entered.

**Comment by bhedayatian** [comment]
I am not able to upgrade to the latest fabric-ui-react ( can't start the app and there is no error messages) I was able to use floating picker from this fabric-ui-react 5.86.0 but I get this error:

MaskedTextField.js:7 Uncaught Error: Cannot find module "@uifabric/utilities/lib/inputMask"
(its intermittent thought)
I thought this was fixed in version  5.86.0???

what's the first stable build for this issue?

**Comment by bhedayatian** [comment]
I am getting the following error using @5.135.1

 Uncaught TypeError: newProps.componentRef is not a function
    at FloatingPeoplePicker.BaseComponent._updateComponentRef (BaseComponent.js:163)
    at FloatingPeoplePicker.BaseComponent.componentDidMount (BaseComponent.js:53)
    at FloatingPeoplePicker.obj.<computed> [as componentDidMount] (BaseComponent.js:217)
    at commitLifeCycles (react-dom.development.js:20049)
    at commitLayoutEffects (react-dom.development.js:22813)
    at HTMLUnknownElement.callCallback (react-dom.development.js:347)



this is the code:
class MyComponent extends BaseComponent {
constructor(props) {
        super(props);
        this._picker = React.createRef();
        this._inputElement = React.createRef();
}


render(){

return(
 <div
          style={{ width: 208, display: "flex", border: "solid black 1px" }}
          ref={this._inputElement}>
          <span>To:</span>
          <input
            placeholder="Search for person"
            style={{ border: "none" }}
            value={this.state.currentValue}
            onChange={this._onSearchChange}
            onFocus={this._onFocus}
          />
        </div>
      <FloatingPeoplePicker
          onResolveSuggestions={this._onFilterChanged}
          getTextFromItem={this._getTextFromItem}
          pickerSuggestionsProps={suggestionProps}
          key="normal"
          onRemoveSuggestion={this._onRemoveSuggestion}
          onValidateInput={this._validateInput}
          componentRef={this._picker}
          onChange={this._onPickerChange}
          inputElement={this._inputElement.current}
          resolveDelay={300}
        />
      </div>

);
}

**Comment by Adjective-Object** [comment]
Looks like there's a mismatch between the expected types of componentRef in [BaseFloatingPickerProps](https://github.com/OfficeDev/office-ui-fabric-react/blob/61ffe7ba8a1f40a61a8335ac9ef9fb6e4dc30fd7/packages/office-ui-fabric-react/src/components/FloatingPicker/BaseFloatingPicker.types.ts#L36) and [BaseComponent](https://github.com/OfficeDev/office-ui-fabric-react/blob/1b9405b3b64839975e16ee1fad04d7868d9e3b99/packages/utilities/src/BaseComponent.ts#L16).

https://github.com/OfficeDev/office-ui-fabric-react/blob/61ffe7ba8a1f40a61a8335ac9ef9fb6e4dc30fd7/packages/office-ui-fabric-react/src/components/FloatingPicker/BaseFloatingPicker.types.ts#L36

https://github.com/OfficeDev/office-ui-fabric-react/blob/1b9405b3b64839975e16ee1fad04d7868d9e3b99/packages/utilities/src/BaseComponent.ts#L16

Is there a reason BaseComponent only supports  callback refs instead of any `React.Ref<T>` type?

**Comment by joschect** [comment]
@Adjective-Object Base component was first created before we started using React.Ref, I think. Plus we don't support 5.0 any more. The current version of [BaseComponent](https://github.com/OfficeDev/office-ui-fabric-react/blob/master/packages/utilities/src/BaseComponent.types.ts#L10-L12) does implement IRefObject.

@bhedayatian I am not sure why you're experience that error. Based on what @Adjective-Object said you might just need to not extend base component. This isn't an issue and if you have further how-to questions please post them to stack overflow.

**Comment by bhedayatian** [comment]
@joschect 
Can you please give more tips how to reuse PeoplePicker styles in my component that uses floatingPicker? I am a newbie in react and not very familiar with typescript.
Thank you

**Comment by Adjective-Object** [comment]
Is this ![image](https://user-images.githubusercontent.com/1174858/63610081-652cb480-c58d-11e9-84c5-92f04178280c.png) what you're asking how to render? (I've usually heard them called pills before).

If so, you could reach into ExtendedSelectedPeopleList to render its pill component in your consumer:
https://github.com/OfficeDev/office-ui-fabric-react/blob/c4a87826fbb9e3d85cba04be9b829d179b1e7af7/packages/office-ui-fabric-react/src/components/SelectedItemsList/SelectedPeopleList/Items/ExtendedSelectedItem.tsx#L16

What's your use case where you're writing a custom item to manage selected items *and* show a picker that isn't one of the picker components in fabric already?

**Comment by bhedayatian** [comment]
@Adjective-Object Yes, I am trying to use the pill component in floatingPicker. I was originally using normalPeople Picker. my use case is about handing multiple email addresses entered by user from copy/paste and I need to access the input DOM itself. @joschect suggested to use floatigpicker ( Indicating we can't access the input DOM in peoplePicker).  the look in floatingPicker is different when user selects an item ( doesn't show it as a pill component and its a plain text)

**Comment by joschect** [comment]
@bhedayatian Per my earlier comment, the initial question has been answered and these additional questions would be better on Stack overflow. If there is a new issue/bug please create a new ticket so we can keep each issue distinct and focused.

**Comment by kldev** [comment]
Sample code: 
`
class XXX extends React.Component<Props> {
    private _picker = React.createRef<IBasePicker<IPersonaProps>>();

    removeFromCode = () => {
     le
t newSelectedItems = [...this._picker.current.items];
   // logic of delete
    // newSelectedItems.filter() some filter function
    this.setState({selectedItems: newSelectedItems })
    }
  }

     <NormalPeoplePicker        
          componentRef={this._picker}   /// assign picker ref
          selectedItems={this.state.selectedItems}
          resolveDelay={300}
        />`

**Comment by xugao** [comment]
@bhedayatian - looks like Jon has answered your question. I am going to close this long-opened issue for now. feel free to re-open if you still have questions
